arcgis-java-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/skins/UtilityNetworkTraceSkin.java

```diff
@@ -0,0 +1,539 @@
+
+    // only show the results tab when there are results and/or a result is in progress
+    noResultsFoundVBox.visibleProperty().bind(Bindings.and(isTraceInProgressProperty.not(), traceResultsProperty.emptyProperty()));
+
+    // keep trace results in sync with data
+    traceResultsProperty.addListener((ListChangeListener<UtilityNetworkTraceOperationResult>) c -> {
+      while (c.next()) {
+        if (traceResultsProperty.isEmpty()) {
+          tabPane.getSelectionModel().select(newTraceTab);
+        }
+        for (UtilityNetworkTraceOperationResult removedResult : c.getRemoved()) {
+          resultsTabPane.getTabs().remove(findTabForResult(removedResult));
+        }
+        for (UtilityNetworkTraceOperationResult addedResult : c.getAddedSubList()) {
+          Platform.runLater(() -> {
+            // waits for result to finish process via async operation
+            var tab = new UtilityNetworkTraceOperationResultView(this, addedResult);
+            resultsTabPane.getTabs().add(tab);
+            resultsTabPane.getSelectionModel().select(tab);
+          });
+        }
+      }
+    });
+
+    // configure the UI for when a trace is running
+    isTraceInProgressProperty.addListener(((observable, oldValue, newValue) -> {
+      if (newValue) {
+        tabPane.getSelectionModel().select(resultsTab);
+        newTraceTab.setDisable(true);
+        resultsVBox.setVisible(false);
+        traceInProgressVBox.setVisible(true);
+      } else {
+        resultsVBox.setVisible(true);
+        traceInProgressVBox.setVisible(false);
+        newTraceTab.setDisable(false);
+      }
+    }));
+
+    // configure the clear results button and only display when there are trace results
+    clearResultsButton.setOnAction(e -> traceResultsProperty.clear());
+    clearResultsButton.visibleProperty().bind(Bindings.isNotEmpty(traceResultsProperty));
+
+    getChildren().add(root);
+  }
+
+  /**
+   * Returns the Tab that is displaying the data for the provided result.
+   *
+   * @return the Tab
+   * @since 100.15.0
+   */
+  private Tab findTabForResult(UtilityNetworkTraceOperationResult result) {
+    return resultsTabPane.getTabs().stream()
+      .filter(c -> ((UtilityNetworkTraceOperationResultView) c).getResult() == result).findFirst().orElse(null);
+  }
+
+  /**
+   * Creates and returns the default trace name used when a name is not defined in the text field.
+   *
+   * @return the default trace name
+   * @since 100.15.0
+   */
+  public String getDefaultTraceName() {
+    return selectedTraceConfigurationProperty.get().getName() + " " + (traceResultsProperty.size() + 1);
+  }
+
+  /**
+   * Handles clicks on the addStartingPointsButton.
+   *
+   * @param mouseEvent the mouse event captured by the click
+   * @since 100.15.0
+   */
+  @FXML
+  private void handleAddStartingPointButtonClicked(MouseEvent mouseEvent) {
+    if (mouseEvent.getButton() == MouseButton.PRIMARY && mouseEvent.isStillSincePress()) {
```

Comment:
Do you need to check for mouse still and what button for pressing a button? I've always done it through setting the `onAction` so could be you are using it in a different way so need to do these checks.

Reply:
No you're right, I think I overlooked the `onAction` tag from FXML - I've tidied that up now. onAction is definitely better for also handling hits on enter when focused etc. for default button behaviour.